### PR TITLE
redis: add memcached proxy for envoy

### PR DESCRIPTION
**Pull request by x-lambda**
redis: add memcached proxy for envoy
 
Commit Message: we expand redis filter to support Redis/Memcached protocol by configurable, default is redis. we add mc codec,
envoy will replace redis-codec with mc-codec, if Memcached protocol configured.
and expand ring hash LB, in order to be compatible with twemproxy, so envoy can completely replace twemproxy.
finally, we introduced twem cluster, which base on eds cluster.
 
Additional Description:
    * why reuse redis filter:
        Like redis, mc is text protocol. And sometimes get multiple keys in one request.
        It is also necessary to select the host through a configurable hash strategy,
        like ring hash/maglev.
 
    * how to support mc cmd:
        We append 'mc' if the cmd and redis cmd are duplicates, the handler cannot be registered also.
        So get cmd is called 'getmc' in mc-codec. And send buffer to upstreams is turn into 'get'.
 
    * how to support noreply:
        Actually, we send buffer to upstream without noreply, when we parse out 'noreply' from client.
        And response received will be ignored from upstream.The reason for this is because only
        the upstream connection has data returned, the downstream connection will be processed,
        otherwise, the downstream connection will be pending.
                   noreply              buffer without noreply
        client----------------->envoy-------------------------->upstream
              <----------------[error] <-----------------------
               decode(ignored)                 response
 
    * why expand RingHash:
        To be compatible with twemproxy. It's different from envoy to generates ring. Twem has custom
        hash func which called ketama, see https://github.com/twitter/twemproxy/blob/master/src/hashkit/nc_ketama.c#L31.
        And use hostname differently, so we extracted this part of the code, with doHash().
 
    * how modify lb:
        We create new class which called 'TwemClusterThreadAwareLoadBalancer', and Inherited from RingHashLoadBalancer.
        We have modified the way to generate rings in createLoadBalancer call.
        Similarly, the key in the request is mapped to the ring has also changed, from crc to fnv.
 
Risk Level: high
Testing: N/A
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional API Considerations:]

### CODE REVIEW

**Comment by wbpcode** [comment]
Great work 👍.
It would be better to create a new proxy for memcached rather then reuse redis proxy.

**Comment by mattklein123** [comment]
Yes per @wbpcode we won't be able to take a patch like this. A few things:
1) Since Redis is stable, it would be better for this to be a new extension. If we want to factor our common code that is fine, but that should be done in independent changes.
2) Any changes to core should be done as independent changes.
3) You will need to find a maintainer sponsor for this extension before proceeding. I'm not sure there is much bandwidth for a new extension like this right now, but I'm not sure. We are trying to actively sort out a "contrib" directory where the bar for new extensions will be lower.

/wait

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 1]
```diff
@@ -57,6 +57,16 @@ message RedisProxy {
       ANY = 4;
     }
 
+    // Protocol controls how Envoy parse client's request. This is currently
+    // supported protocols are memcached and redis.
+    enum Protocol {
+      // Default mode. Parse for redis.
+      Redis = 0 [(udpa.annotations.enum_value_migrate).rename = "REDIS"];
+
+      // Parse for memcached.
```
Comment (htuch) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:66]: Can you link to how Redis treats memcached requests? Some documentation would help.

**Comment by htuch** [comment]
Tagging @msukalski @henryyyang @mattklein123 as CODEOWNERS

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto (hunk 1 above)
Comment (htuch) [reply api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:66]: Otherwise LGTM for API.

**Comment by lvht** [comment]
Hi @htuch @mattklein123 @wbpcode , this is an co-author of this PR.  Sorry for late reply.

Here here reasons why to reuse the redis proxy code.

It is wired to reuse the redis proxy's code. However, if we implement the memecached proxy in an independent extension, you will find that we re-implement too many features. This is why @maximebedard made [this PR](https://github.com/envoyproxy/envoy/pull/6800).

Almost all the features of the non-clustered redis proxy could be reused by memcached proxy. 

The filter, router, command spliter and client will be reimplemented, with almost the identical feature.

We have tried to split these common code into a separate place. However, the router and command spliter a too tight with redis code to split.

Oh the other hand, if we don't insist memcached should have its own extension, we can implement it very easily.

mcmcached protocol is very similar to redis protocol. Both are text protocol, both are line based, both are distributed by key, and both need split by mulitiple keys.

Almost every code needed by memcached have been offered by redis.

The most important port of redis code that memcached can reuse is the command spliter.

We only add 468 line the `source/extensions/filters/network/common/redis` to support memcached protocol. And most of these lines(351 line) are memcached codec.

For memecached cmd "cas", "add", "replace", "prepend", "delete", "touch", we just added them to the simpleCommands list. For "get" and "gets", we create the McRetrievalRequest, which is based on the FragmentedRequest, as well.

Memcached protocol has some different with redis. There are two issues we need to address.

One issue is the command collision. Both memcached and redis use the command "get" and "touch". We need an way to distinct them. So we decide to append a "mc" suffix to memcached "get" and "touch" while decoding and trim such suffix when encoding. So you will see the "getmc" and "touchmc" commands in the code base. It is ugly, but it works, and works without too much mantiance burden.

The other thing is the noreply option. It seems that the redis filter assume every upstream request will get a reply. And this is not true for memcached. If you offer an noreply option, memcached will not reply the response. However, if error occured during the server decoding request, the client will see an error reply. So we decide to drop the noreply option for the upstream request. In other word, we let enovy set request to upstream without noreply option, so we will always get a reply. If envoy need to send reply to the downside, it will check the noreply option. So we need to store the noreply option. This is why we change the `RespValue` class and add the noreply_ field. Every request is a RespValue.

By reuse the redis code, we could make as little as possible change, and introduce as littler maintain burden as possible.

Please  make an early review the the codec implement, before made decision. Thank you.

**Comment by htuch** [comment]
We could go either way here; what's important is that we have existing CODEOWNERS comfortable with the change (CC @wbpcode @msukalski) or create a new extension and  start by finding a maintainer who will sponsor the extension as per https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#adding-new-extensions.

I agree with the goal of DRY and sharing code; this could be a library or we can follow the same extension approach.

This is a large PR, once we have a review strategy, I think we need to break out pieces into separate PRs, e.g. the ring hash LB changes.

**Comment by lvht** [comment]
@htuch we could break out this large PR into seprate PRS if needed.

**Comment by mattklein123** [comment]
Yeah my main concern here is not the design (sharing), it's more that I don't think we have any active redis code owners right now and I'm not sure where the review bandwidth is going to come from realistically. I can potentially help with this but not until June as I'm mostly OOO until then.

**Comment by HenryYYang** [comment]
I totally understand the desire to reuse the code, but I think there're more features in Redis that's not currently supported by the Redis filter, by trying to support both memcached and redis, it'd actually make it much more difficult to continue to improve the Redis proxy.  Some of the big one are the streaming interface or some of the commands that require dedicated connection for a particular client.

**Pull request closed without merging**